### PR TITLE
assert: add compile-time option to allow logging debug assertion failures in release builds

### PR DESCRIPTION
**Pull request by ggreenway**
Signed-off-by: Greg Greenway <ggreenway@apple.com>

*Risk Level*: Low
*Testing*: Added new tests, augmented existing tests, added to CI
*Docs Changes*: added
*Release Notes*:

### CODE REVIEW

File: source/common/common/assert.h [hunk 1]
```diff
@@ -1,8 +1,54 @@
...
+  virtual ~ActionRegistration() {}
+};
+typedef std::unique_ptr<ActionRegistration> ActionRegistrationPtr;
+
+/**
+ * Sets an action to be invoked when a debug assertion failure is detected
+ * in a release build. This action will be invoked each time an assertion
+ * failure is detected.
+ *
+ * This function is not thread-safe; concurrent calls are not allowed.
```
Comment (mattklein123) [inline source/common/common/assert.h]: FYI there is no guarantee that asserts will not fire concurrently in the current code. I'm not sure it matters with what you are doing in your invoked function, but do you want to do something about this or change the comment here?

File: source/common/common/assert.cc [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    if (debugAssertionFailureRecordAction_ != nullptr) {
+      debugAssertionFailureRecordAction_();
+    }
+  }
+
+private:
+  // This implementation currently only handles one action being set at a time. This is currently
+  // sufficient. If multiple actions are ever needed, the actions should be chained when
+  // additional actions are registered.
+  static std::function<void()> debugAssertionFailureRecordAction_;
```
Comment (mattklein123) [inline source/common/common/assert.cc]: nit: snake case

File: source/common/common/assert.h [hunk 3]
```diff
@@ -1,8 +1,54 @@
...
+ */
+ActionRegistrationPtr setDebugAssertionFailureRecordAction(std::function<void()> action);
+
+/**
+ * Invokes the action set by setDebugAssertionFailureRecordAction, or does nothing if
+ * no action has been set.
+ *
+ * This should only be called by ASSERT macros in this file.
+ */
+void invokeDebugAssertionFailureRecordAction_ForAssertMacroUseOnly();
```
Comment (mattklein123) [inline source/common/common/assert.h]: For my edification why is this indirection needed?

File: source/common/common/assert.cc [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    ASSERT(debugAssertionFailureRecordAction_ == nullptr);
+    debugAssertionFailureRecordAction_ = action;
+  }
+
+  ~ActionRegistrationImpl() {
+    ASSERT(debugAssertionFailureRecordAction_ != nullptr);
+    debugAssertionFailureRecordAction_ = nullptr;
+  }
+
+  static void invokeAction() {
```
Comment (mattklein123) [inline source/common/common/assert.cc]: Technically if we started other threads before we register the callback, we might have threading issues here. This is somewhat related to my other comment. The way you actually use this, I don't think it matters, but should we add some more comments on current usage and limitations?

File: source/common/common/assert.h (hunk 1 above)
Comment (ggreenway) [reply source/common/common/assert.h]: I'll clarify the comment; what I mean here is that it isn't safe to concurrently set the assert callback, not that the assert callback cannot be called concurrently.

File: source/common/common/assert.h (hunk 3 above)
Comment (ggreenway) [reply source/common/common/assert.h]: For a little bit cleaner encapsulation, and it probably keeps the emitted code a bit smaller. But the stored value could be extern'd here, and that would work too.

File: source/common/common/assert.cc (hunk 4 above)
Comment (ggreenway) [reply source/common/common/assert.cc]: Adding more comments.

**Pull request merged**